src/core/attribution/aggregation.js

```diff
@@ -25,11 +30,129 @@ export type ConnectionAggregation = {|
   +nodeAggregations: $ReadOnlyArray<NodeAggregation>,
 |};
 
+function sumScoredConnections(xs: $ReadOnlyArray<ScoredConnection>): number {
+  const reduce = (totalScore, scoredConnection) =>
+    scoredConnection.connectionScore + totalScore;
+  return xs.reduce(reduce, 0);
+}
+
+/* The correctness of this function depends on the following condition: For
+ * every ScoredConnection in xs, there is exactly one NodePrefix for which the
+ * connection's source has that NodePrefix as a prefix.  When the condition is
+ * not met, the function will almost always error, but in certain cases it can
+ * return incorrect results (example in test code).
+ */
+export function aggregateByNodePrefix(
+  xs: $ReadOnlyArray<ScoredConnection>,
+  nodePrefixes: $ReadOnlyArray<NodeAddressT>
+): $ReadOnlyArray<NodeAggregation> {
+  let connectionsFound = 0;
+  const result = [];
+  for (const nodePrefix of nodePrefixes) {
+    const connections = xs.filter(({source}) =>
+      NodeAddress.hasPrefix(source, nodePrefix)
+    );
+    const size = connections.length;
+    connectionsFound += size;
+    const aggregateScore = sumScoredConnections(connections);
+    const aggregation: NodeAggregation = {
+      nodePrefix,
+      connections,
+      aggregationSummary: {size, aggregateScore},
+    };
+    result.push(aggregation);
+  }
+  if (connectionsFound !== xs.length) {
+    // It would be extremely misleading to silently return an aggregation that
+    // did not successfully aggregate every connection, as this would lead to
+    // pieces of the graph silently disappearing.
+    throw new Error(
+      "Invariant violation: connections found is not same as connections in"
+    );
+  }
+  return result;
+}
+
+/* The correctness of this function depends on the following conditions:
```

Comment:
One unfortunate property of this implementation is that in the case that
we add new kinds of connection types, Flow will not direct us to modify
this function, and the connections will be silently omitted from the
aggregation. Do you have any ideas about how we might improve this?

Reply:
Fixed in the rewrite.